prost_build/src/main.rs

```diff
@@ -0,0 +1,155 @@
+                .filter(|e| {
+                    e.file_type().is_file()
+                        && e.path().extension().is_some()
+                        && e.path().extension().unwrap() == "proto"
+                })
+                .map(|e| e.into_path())
+                .collect(),
+        );
+    }
+
+    // create directories for temporary build dirs
+    fs::create_dir_all(tmp_dir)
+        .unwrap_or_else(|_| panic!("Failed to create {:?}", tmp_dir.to_str()));
+
+    // Compile all proto files
+    let mut config = prost_build::Config::default();
+    config.out_dir(tmp_dir);
+    config
+        .compile_protos(&protos, &proto_include_paths)
+        .unwrap();
+
+    // Compile all proto client for GRPC services
+    println!("[info ] Compiling proto clients for GRPC services!");
+    tonic_build::configure()
+        .build_client(true)
+        .build_server(false)
+        .format(true)
+        .out_dir(tmp_dir)
+        .compile(&protos, &proto_include_paths)
+        .unwrap();
+
+    copy_generated_files(tmp_dir, out_dir);
+
+    println!("[info ] => Done!");
+}
+
+fn copy_generated_files(from_dir: &Path, to_dir: &Path) {
+    println!("Copying generated files into '{}'...", to_dir.display());
+
+    // Remove old compiled files
+    remove_dir_all(&to_dir).unwrap_or_default();
+    create_dir_all(&to_dir).unwrap();
+
+    let mut filenames = Vec::new();
+
+    // Copy new compiled files (prost does not use folder structures)
+    let errors = WalkDir::new(from_dir)
+        .into_iter()
+        .filter_map(|e| e.ok())
+        .filter(|e| e.file_type().is_file())
+        .map(|e| {
+            let filename = e.file_name().to_os_string().to_str().unwrap().to_string();
+            filenames.push(filename.clone());
+            copy_and_patch(e.path(), format!("{}/{}", to_dir.display(), &filename))
+        })
+        .filter_map(|e| e.err())
+        .collect::<Vec<_>>();
+
+    if !errors.is_empty() {
+        for e in errors {
+            eprintln!("[error] Error while copying compiled file: {}", e);
+        }
+
+        panic!("[error] Aborted.");
+    }
+}
+
+fn copy_and_patch(src: impl AsRef<Path>, dest: impl AsRef<Path>) -> io::Result<()> {
+    // Skip proto files belonging to `EXCLUDED_PROTO_PACKAGES`
+    for package in EXCLUDED_PROTO_PACKAGES {
+        if let Some(filename) = src.as_ref().file_name().and_then(OsStr::to_str) {
+            if filename.starts_with(&format!("{}.", package)) {
+                return Ok(());
+            }
+        }
+    }
+
+    let contents = fs::read_to_string(src)?;
+
+    // `prost-build` output references types from `tendermint-proto` crate
+    // relative paths, which we need to munge into `tendermint_proto` in
+    // order to leverage types from the upstream crate.
+    let contents = Regex::new(COSMOS_SDK_PROTO_REGEX)
+        .unwrap()
+        .replace_all(&contents, "cosmos_sdk_proto::cosmos");
+
+    fs::write(dest, contents.as_bytes())
+}
```

Comment:
this is a port of the first Rust code I ever read and am glad to see it makes 1000x more sense now